### PR TITLE
[RDY] eval, ex_getln: Fix incompatible pointer types

### LINKED ISSUE #8786: compiler warning: cast between incompatible function types (ExpandUserDefined)
```
[76/182] Building C object src/nvim/CMakeFiles/nvim.dir/ex_getln.c.o
../src/nvim/ex_getln.c: In function ‘ExpandUserDefined’:
../src/nvim/ex_getln.c:5071:34: warning: cast between incompatible function types from ‘char * (*)(const char * const,  const int,  const char_u * const* const,  const _Bool)’ {aka ‘char * (*)(const char * const,  const int,  const unsigned char * const* const,  const _Bool)’} to ‘void * (*)(char_u *, int,  char_u **, int)’ {aka ‘void * (*)(unsigned char *, int,  unsigned char **, int)’} [-Wcast-function-type]
   retstr = call_user_expand_func((user_expand_func_T)call_func_retstr, xp,
                                  ^
```

The code there appears to be older already (28dafe3ff, by @ZyX-I).

gcc (GCC) 8.1.1 20180531

### PR DESCRIPTION
**Pull request by gyatskov**
Fixes #8786

### CODE REVIEW

File: src/nvim/eval.c [hunk 1]
```diff
@@ -1186,7 +1186,7 @@ int call_vim_function(
     const char_u *func,
     int argc,
     const char_u *const *const argv,
-    int safe,                       // use the sandbox
+    _Bool safe,                       // use the sandbox
```
Comment (bfredl) [inline src/nvim/eval.c]: use `bool`

File: src/nvim/ex_getln.c [hunk 2]
```diff
@@ -201,7 +201,7 @@ static Array cmdline_block = ARRAY_DICT_INIT;
 /*
  * Type used by call_user_expand_func
  */
-typedef void *(*user_expand_func_T)(char_u *, int, char_u **, int);
+typedef void *(*user_expand_func_T)(const char_u *, const int, const char_u * const * const, const _Bool);
```
Comment (bfredl) [inline src/nvim/ex_getln.c]: Only `const` behind a pointer is meaningful in a signature, so: `(const char_u *, int, const char_u *const *, bool)`

File: src/nvim/eval.c (hunk 1 above)
Comment (justinmk) [reply src/nvim/eval.c]: @hyatskov this `_Bool` still needs to be changed.

 Also the PR has conflicts, needs a rebase.

File: src/nvim/ex_getln.c [hunk 3]
```diff
@@ -5046,7 +5049,10 @@ static void * call_user_expand_func(user_expand_func_T user_expand_func,
   ccline.cmdprompt = NULL;
   current_SID = xp->xp_scriptID;
 
-  ret = user_expand_func(xp->xp_arg, 3, args, FALSE);
+  ret = user_expand_func(xp->xp_arg,
+                         3,
+                         (const unsigned char * const *)args,
+                         false);
```
Comment (justinmk) [inline src/nvim/ex_getln.c]: `char_u` instead of `unsigned char`

File: src/nvim/eval.c (hunk 1 above)
Comment (justinmk) [reply src/nvim/eval.c]: `_Bool` needs to be changed to `bool`.

Comment (hyatskov) [reply src/nvim/eval.c]: My bad.

File: src/nvim/ex_getln.c (hunk 3 above)
Comment (hyatskov) [reply src/nvim/ex_getln.c]: Done

File: src/nvim/eval.c [hunk 4]
```diff
@@ -1186,7 +1186,7 @@ int call_vim_function(
     const char_u *func,
     int argc,
     const char_u *const *const argv,
-    int safe,                       // use the sandbox
+    bool safe,                       // use the sandbox
     int str_arg_only,               // all arguments are strings
```
Comment (justinmk) [inline src/nvim/eval.c]: small nit: if comments are aligned, try to keep them aligned.

**Comment by janlazo** [comment]
No one checked if 8.1.0053 is relevant?

**Comment by justinmk** [comment]
@janlazo no

**Pull request merged**